Python/Algorithms/data_structures/binary_tree.py

```diff
@@ -0,0 +1,214 @@
+    
+    def create_root_if_none(self,value):
+        if self.root is None:
+            self.root = self.make_node(value)
+            self.length += 1
+            return True
+        return False
+    
+    def make_node(self, value):
+        if value is None:
+            raise ValueError("Nonetype Argument not expected")
+        
+        return value if isinstance(value, TreeNode) else TreeNode(value)
+
+    def check_remove_root(self):
+        '''
+        Helper function to check and remove if it has no tree nodes or 1 tree node.
+        '''
+        if self.root is None:
+            return True
+        if self.root.next is None:
+            self.root = None
+            self.length -= 1
+            return True
+
+        return False
+    
+    def inorder(self, node = None, fn = print):
+        '''
+        Applies `fn` to all elements during inorder traversal
+
+        '''
+        if node is None:
+            node = self.root
+        if node.left:
+            self.inorder(node.left,fn)
+
+        fn(node.value)
+        
+        if node.right:
+            self.inorder(node.right,fn)
+
+    def preorder(self, node = None, fn = print):
+        '''
+        Applies `fn` to all elements during preorder traversal
+
+        '''
+        if node is None:
+            node = self.root
+        
+        fn(node.value)
+        if node.left:
+            self.preorder(node.left,fn)
+        if node.right:
+            self.preorder(node.right,fn)
+    
+    def postorder(self, node = None, fn = print):
+        '''
+        Applies `fn` to all elements during postorder traversal
+
+        '''
+        if node is None:
+            node = self.root
+        
+        if node.left:
+            self.postorder(node.left,fn)
+
+        if node.right:
+            self.postorder(node.right,fn)
+
+        fn(node.value)
+    
+    def levelorder(self, node = None, fn = print):
+        '''
+        Applies `fn` to all elements during level order traversal
+        '''
+        node = node if node else self.root
+        h = self.height(node)
+        for i in range(1,h+1):
+            self.apply_to_level(node, i, print)
+
+    def apply_to_level(self, node, level, fn = print):
+        '''
+        Applies given function to a level.
+        '''
+        if node is None:
+            return
+        if level == 1:
+            fn(node.value)
+        if level > 1:
+            self.apply_to_level(node.left, level - 1, fn)
+            self.apply_to_level(node.right, level - 1, fn)
+
+    
+    def height(self, node):
+        '''
+        find the height of a node in the tree from the bottom most element in that subtree.
+
+        '''
+        if node is None:
+            return 0
+        
+        lheight = self.height(node.left)
+        rheight = self.height(node.right)
+
+        if lheight > rheight:
+            return lheight + 1
+        else:
+            return rheight + 1
+
+    def dfs(self, v):
+        '''
```

Comment:
That's right. A graph data structure will be a better generalization for a tree which can have more than two children. There we'll probably use something like adjacency lists, BFS using adjacency lists will be applicable for such trees. But here in the case of binary trees, instead of adjacency lists we are using `left` and `right` attributes, this structure is different than the generic graph structure. I think that will be a different implementation?